src/Microsoft.OData.Core/UriParser/SemanticAst/SinglePropertyCastNode.cs

```diff
@@ -0,0 +1,75 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="SinglePropertyCastNode.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using Microsoft.OData.Edm;
+
+namespace Microsoft.OData.UriParser
+{
+    /// <summary>
+    /// Node representing a type segment that casts a single entity/complex parent node.
+    /// </summary>
+    public sealed class SinglePropertyCastNode : SingleValueNode
+    {
+        /// <summary>
+        /// The resource that we're casting to a different type.
+        /// </summary>
+        private readonly SingleValueNode source;
+
+        /// <summary>
+        /// The target type that the source is cast to.
+        /// </summary>
+        private readonly IEdmPrimitiveTypeReference primitiveTypeReference;
+
+        /// <summary>
+        /// Created a SinglePropertyCastNode with the given source node and the given type to cast to.
+        /// </summary>
+        /// <param name="source"> Source <see cref="SingleValueNode"/> that is being cast.</param>
+        /// <param name="primitiveType">Type to cast to.</param>
+        /// <exception cref="System.ArgumentNullException">Throws if the input entityType is null.</exception>
+        public SinglePropertyCastNode(SingleValueNode source, IEdmPrimitiveType primitiveType)

```

Comment:
>source [](start = 54, length = 6)

Can source be null? If not, should we add an not null check here? I notice that we don't on SingleResourceCastNode, which also seems strange.